Add option to update product description

anitaa1990 opened the pull request:
Fixes Step 3 of #1594. This PR adds the ability to update the product description.

### Notes
- There is no option right now to update an **empty** product description at the moment. This will need a fix in FluxC and will be taken care of in another PR.
- There are TODO's in the code to add tracking events which will be added in another PR.
- The changes in this PR are only available to `DEBUG` users.

### Screenshots:
(LEFT: displays existing product description . RIGHT: displays when there is no product description)
<img width="300" src="https://user-images.githubusercontent.com/22608780/69845358-6eb56f00-1296-11ea-90fb-b10bbbd38e0d.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/69845362-707f3280-1296-11ea-8a06-55e78dd86c9e.png">

### Behaviour
(LEFT: Update product description . RIGHT: discard changes)
<img width="300" src="https://user-images.githubusercontent.com/22608780/69846950-ad015d00-129b-11ea-8b10-cc49469b10da.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/69846955-aecb2080-129b-11ea-851c-25d3b4af1731.gif">

### Testing
- Adding description to a product.
- Updating description of a product.
- Verify that images can be added added to the product description using the HTML editor.
- Able to remove images from the description.
- Formatting changes in the description (adding/removing lists/headings/links) are updated correctly.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nbradbury commented:
@anitaa1990 I'm getting this exception when I switch to the product list:

```
  java.lang.IllegalArgumentException: Date string argument is not of format yyyy-MM-dd'T'HH:mm:ss: 
        at com.woocommerce.android.extensions.StringDateFormatExtKt.dateFromIso8601(StringDateFormatExt.kt:132)
        at com.woocommerce.android.model.ProductKt.toAppModel(Product.kt:79)
```

nbradbury commented:
Typo here - "sace" instead of "save."

![Screenshot_1575289509](https://user-images.githubusercontent.com/3903757/69959353-f688bc80-14d4-11ea-8380-54597471d005.png)

nbradbury commented:
@anitaa1990 If you press Done in the editor, the description is updated in the product detail. But if you back out of product detail without tapping Update, there's no warning that changes will be lost.

Also, if you change some simple formatting - like bolding a word - those changes aren't reflected in the detail. I think we can use the HTML features of `TextView` to show simple formatting in the detail screen. @kyleaparker Sound okay to you?

nbradbury commented:
If I try to update a description of a product that's a variation, I can this error response:

```
{"error":"woocommerce_rest_invalid_product_id","message":"To manipulate product variations you should use the \/products\/&lt;product_id&gt;\/variations\/&lt;id&gt; endpoint."}
```

nbradbury commented:
> Verify that images can be added added to the product description using the HTML editor.

I'm not seeing a way to add images. How do I do this?

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/aztec/AztecEditorFragment.kt:
```diff
@@ -1,29 +1,45 @@
+class AztecEditorFragment : BaseFragment(), IAztecToolbarClickListener, BackPressListener {
     companion object {
         const val TAG: String = "AztecEditorFragment"
+        const val AZTEC_EDITOR_REQUEST_CODE = 3001
```
At some point we'll want to consolidate all the app's request codes to avoid conflicts. I've opened [a separate issue](https://github.com/woocommerce/woocommerce-android/issues/1618) to address this.

anitaa1990 replied:
Makes sense! Thanks for adding the issue 👍

nbradbury commented:
In WPAndroid (left) the toolbar has a "..." item on the right to indicate there are more toolbuttons, but in WCAndroid (right) this "more" item isn't shown. This was added to WPAndroid because people weren't aware the toolbar could be scrolled. Is this something we can show in WCAndroid, too?

![wpa](https://user-images.githubusercontent.com/3903757/69970572-7706e780-14ec-11ea-9491-c05a64bc9a60.png)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/aztec/AztecEditorFragment.kt:
```diff
@@ -46,15 +62,78 @@ class AztecEditorFragment : BaseFragment(), IAztecToolbarClickListener {
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        outState.putBoolean(FIELD_IS_CONFIRMING_DISCARD, isConfirmingDiscard)
```
We should also save `isHtmlEditorEnabled` here. If the activity is recreated the Aztec editor correctly switches to HTML mode, but `isHtmlEditorEnabled` will be `false`, which will cause problems in [getEditorText](https://github.com/woocommerce/woocommerce-android/pull/1613/files#diff-f835bd0be0106f3de0fea57bcd7430b7R166).

anitaa1990 replied:
Good point! Fixed in dd2a2dd

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -564,6 +608,14 @@ class ProductDetailFragment : BaseFragment(), OnGalleryImageClickListener {
+    override fun onNavigationResult(requestCode: Int, result: Bundle) {
+        when (requestCode) {
+            AZTEC_EDITOR_REQUEST_CODE -> {
+                viewModel.updateProductDraft(result.getString(ARG_AZTEC_EDITOR_TEXT))
```
We should include a boolean in the `result` bundle which tells us whether the user made changes. Otherwise the "Update" menu will appear if the user hits "Done" but made no changes.

anitaa1990 replied:
👍 Done in 42f3e73

nbradbury commented:
@anitaa1990 Done with my first round!

kyleaparker commented:
>Also, if you change some simple formatting - like bolding a word - those changes aren't reflected in the detail. I think we can use the HTML features of TextView to show simple formatting in the detail screen. @kyleaparker Sound okay to you?

👍

anitaa1990 commented:
Thanks for the review @nbradbury!

> The buttons don't really make sense to me. The dialog is asking whether to save the changes, but the buttons say "Cancel" and "Discard." I think the message should be something like "Discard changes?" and the buttons should be "Discard" and "Don't Discard" (or "Discard" and "Cancel").

Fixed this in 42f3e73  as per design specified:
<img width="265" src="https://user-images.githubusercontent.com/22608780/70062540-38684e80-160c-11ea-9d09-2844e009bb53.png"> .  <img width="265" src="https://user-images.githubusercontent.com/22608780/70062609-5b92fe00-160c-11ea-9755-fabf0d47b778.png">

> f you press Done in the editor, the description is updated in the product detail. But if you back out of product detail without tapping Update, there's no warning that changes will be lost.

I was meaning to do that and it completely slipped my mind 🤦‍♀ ! Fixed in e211a92

> Also, if you change some simple formatting - like bolding a word - those changes aren't reflected in the detail. I think we can use the HTML features of TextView to show simple formatting in the detail screen

👍 Done in 365de87

> If I try to update a description of a product that's a variation, I see this error response:
```{"error":"woocommerce_rest_invalid_product_id","message":"To manipulate product variations you should use the \/products\/&lt;product_id&gt;\/variations\/&lt;id&gt; endpoint."}```

I'm not able to reproduce this on my test site. Is it happening for all variable products? Could you provide me access to your test site so I can debug what's going on here?

> Verify that images can be added added to the product description using the HTML editor. 
I'm not seeing a way to add images. How do I do this?

Sorry, I should have been more clear! I meant we would need to add an image using `wp-admin` and verify that the image is being displayed in the Aztec editor in the app. And when updating product description, the existing image should not be affected. And removing the image from the editor should also be reflected.

> In WPAndroid (left) the toolbar has a "..." item on the right to indicate there are more toolbuttons, but in WCAndroid (right) this "more" item isn't shown. This was added to WPAndroid because people weren't aware the toolbar could be scrolled. Is this something we can show in WCAndroid, too?

Yeah the design also specifies the same thing! But this toolbar with the "more" item is initialised only when the media option is supported. I remember debugging a crash that happened in a [previous PR](https://github.com/woocommerce/woocommerce-android/pull/1553#issuecomment-556020575) and removed this option for now. We can probably add it when we start supporting media or maybe add a fix in Aztec-Android, if it's relatively straightforward (I must confess I haven't actually looked into the fix yet).


Ready for another round 👍

nbradbury commented:
> I'm not able to reproduce this on my test site. Is it happening for all variable products?

I just tried it again on the very same product, and now it works. Sounds like we have another mystery! It might be worthwhile asking Proton about that error message.

> But this toolbar with the "more" item is initialised only when the media option is supported

👍 

> Ready for another round 

On it!

nbradbury commented:
Changes look good, but the discard confirmations are different for the product detail screen and Aztec screen. I think we'd want to use the same wording & buttons in both places.

![aztec](https://user-images.githubusercontent.com/3903757/70070504-b6ebce80-15c1-11ea-9b36-73c590cb399e.png)
![detail](https://user-images.githubusercontent.com/3903757/70070505-b6ebce80-15c1-11ea-9fd6-ff3de7eabcda.png)

anitaa1990 commented:
Thanks @nbradbury!

> The discard confirmations are different for the product detail screen and Aztec screen. I think we'd want to use the same wording & buttons in both places.

Modified in 06966f1! Ready for another round 👍
